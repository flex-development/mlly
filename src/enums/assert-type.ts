/**
 * @file Enums - AssertType
 * @module mlly/enums/AssertType
 */

/**
 * `import` assertion types.
 *
 * ::: info
 * It's unclear whether or not the HTML spec will require assertion types for
 * WebAssembly modules.
 * :::
 *
 * @see https://github.com/WebAssembly/esm-integration/issues/42
 *
 * @enum {Lowercase<string>}
 */
enum AssertType {
  IMPLICIT = 'javascript',
  JSON = 'json'
}

export default AssertType
