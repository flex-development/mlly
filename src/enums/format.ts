/**
 * @file Enums - Format
 * @module mlly/enums/Format
 */

/**
 * Module formats.
 *
 * @enum {Lowercase<string>}
 */
enum Format {
  BUILTIN = 'builtin',
  COMMONJS = 'commonjs',
  JSON = 'json',
  MODULE = 'module',
  WASM = 'wasm'
}

export default Format
