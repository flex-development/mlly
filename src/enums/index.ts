/**
 * @file Entry Point - Enums
 * @module mlly/enums
 */

export { default as Format } from './format'
export { default as SpecifierKind } from './kind-specifier'
export { default as StatementKind } from './kind-statement'
export { default as SyntaxKind } from './kind-syntax'
