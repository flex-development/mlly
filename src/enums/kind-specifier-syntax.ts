/**
 * @file Enums - SpecifierSyntaxKind
 * @module mlly/enums/SpecifierSyntaxKind
 */

/**
 * Module specifier syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SpecifierSyntaxKind {
  DYNAMIC = 'dynamic',
  STATIC = 'static'
}

export default SpecifierSyntaxKind
