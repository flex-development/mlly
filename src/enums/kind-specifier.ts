/**
 * @file Enums - SpecifierKind
 * @module mlly/enums/SpecifierKind
 */

/**
 * Module specifier types.
 *
 * @see https://nodejs.org/api/esm.html#terminology
 *
 * @enum {Lowercase<string>}
 */
enum SpecifierKind {
  ABSOLUTE = 'absolute',
  BARE = 'bare',
  RELATIVE = 'relative'
}

export default SpecifierKind
