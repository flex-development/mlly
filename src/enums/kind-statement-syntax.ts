/**
 * @file Enums - StatementSyntaxKind
 * @module mlly/enums/StatementSyntaxKind
 */

/**
 * CommonJS and ESM statement syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum StatementSyntaxKind {
  DECLARATION = 'declaration',
  DEFAULT = 'default',
  DEFAULT_WITH_NAMED = 'default-with-named',
  DEFAULT_WITH_NAMESPACE = 'default-with-namespace',
  DYNAMIC = 'dynamic',
  LIST = 'list',
  NAMED = 'named',
  NAMESPACE = 'namespace',
  REQUIRE = 'require',
  SIDE_EFFECT = 'side-effect'
}

export default StatementSyntaxKind
