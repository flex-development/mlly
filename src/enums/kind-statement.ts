/**
 * @file Enums - StatementKind
 * @module mlly/enums/StatementKind
 */

/**
 * CommonJS and ESM statement kinds.
 *
 * @enum {Lowercase<string>}
 */
enum StatementKind {
  EXPORT = 'export',
  IMPORT = 'import',
  REQUIRE = 'require'
}

export default StatementKind
