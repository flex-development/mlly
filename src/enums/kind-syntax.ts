/**
 * @file Enums - SyntaxKind
 * @module mlly/enums/SyntaxKind
 */

/**
 * CommonJS and ESM statement syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SyntaxKind {
  DECLARATION = 'declaration',
  DEFAULT = 'default',
  DEFAULT_WITH_NAMED = 'default-with-named',
  DEFAULT_WITH_NAMESPACE = 'default-with-namespace',
  DYNAMIC = 'dynamic',
  LIST = 'list',
  NAMED = 'named',
  NAMESPACE = 'namespace',
  REQUIRE = 'require',
  SIDE_EFFECT = 'side-effect'
}

export default SyntaxKind
