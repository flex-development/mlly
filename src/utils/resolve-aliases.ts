/**
 * @file resolveAliases
 * @module mlly/utils/resolveAliases
 */

import { SpecifierSyntaxKind } from '#src/enums'
import type { ResolveAliasOptions } from '#src/interfaces'
import regexp from '#src/internal/escape-reg-exp'
import extractStatements from './extract-statements'
import resolveAlias from './resolve-alias'

/**
 * Resolves path aliases in the given piece of source `code`.
 *
 * @see {@linkcode ResolveAliasOptions}
 * @see {@linkcode resolveAlias}
 *
 * @async
 *
 * @param {string} code - Code to evaluate
 * @param {ResolveAliasOptions} [options={}] - Resolve alias options
 * @return {Promise<string>} `code` unmodified or with path aliases resolved
 */
const resolveAliases = async (
  code: string,
  options: ResolveAliasOptions = {}
): Promise<string> => {
  for (const statement of extractStatements(code)) {
    // do nothing if statement does not have specifier
    if (!statement.specifier) continue

    // ignore statements with dynamic specifiers
    if (statement.specifier_syntax === SpecifierSyntaxKind.DYNAMIC) continue

    /**
     * Resolved module specifier.
     *
     * @const {string} specifier
     */
    const specifier: string = await resolveAlias(statement.specifier, options)

    // do nothing if original specifier did not contain path alias
    if (specifier === statement.specifier) continue

    // replace path alias
    code = code.replace(
      statement.code,
      statement.code.replace(
        new RegExp(`(?<=["'])${regexp(statement.specifier)}(?=["'])`),
        specifier
      )
    )
  }

  return code
}

export default resolveAliases
