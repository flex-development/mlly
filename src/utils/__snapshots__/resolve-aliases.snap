// Vitest Snapshot v1

exports[`unit:utils/resolveAliases > should return code with path aliases resolved 1`] = `
"import type { ParsedModuleId, ResolveAliasOptions } from '../interfaces'
import regexp from '../internal/escape-reg-exp'
import validateBoolean from '../internal/validate-boolean'
import validateObject from '../internal/validate-object'
import validateSet from '../internal/validate-set'
import validateString from '../internal/validate-string'
import validateURLString from '../internal/validate-url-string'
import type { NodeError } from '@flex-development/errnode'
import pathe from '@flex-development/pathe'
import {
  CompareResult,
  isNIL,
  type Nullable
} from '@flex-development/tutils'
import { URL, pathToFileURL } from 'node:url'
import compareSubpaths from './compare-subpaths'
import CONDITIONS from './conditions'
import parseModuleId from './parse-module-id'
import PATTERN_CHARACTER from './pattern-character'
import RESOLVE_EXTENSIONS from './resolve-extensions'
import resolveModule from './resolve-module'
import toBareSpecifier from './to-bare-specifier'
import toRelativeSpecifier from './to-relative-specifier'
import toURL from './to-url'

await import(foo)

export default {}"
`;
